torch/ao/sparsity/sparsifier.py

```diff
@@ -0,0 +1,213 @@
+    return model
+
+
+class BaseSparsifier(abc.ABC):
+    r"""Base class for all sparsifiers.
+
+    Abstract methods that need to be implemented:
+
+    - step: Function to compute a new mask for all keys in the `module_groups`.
+
+    Args:
+        - model [nn.Module]: model to configure. The model itself is not saved
+            but used for the state_dict saving / loading.
+        - config [list]: configuration elements could either be instances of
+            nn.Module or dict maps. The dicts must have a key 'module' with the
+            value being an instance of a nn.Module.
+        - defaults [dict]: default configurations will be attached to the
+            configuration. Only the keys that don't exist in the `config` will
+            be updated.
+
+    Example::
+
+        >>> config = [model.layer1, {'module': model.linear2, 'sparsity_level': 0.5}]
+        >>> defaults = {'sparsity_level': 0.7}
+        >>> # model.layer1 will have `sparsity_level` = 0.7 (getting default)
+        >>> sparsifier = BaseSparsifier(config, defaults)
+    """
+    def __init__(self, defaults):
+        super().__init__()
+        self.defaults = defaults
+        if self.defaults is None:
+            self.defaults = dict()
+
+        self.state = dict()
+        self.module_groups = []
+        self.enable_mask_update = True
+
+    def __getstate__(self):
+        return {
+            'defaults': self.defaults,
+            'state': self.state,
+            'module_groups': self.module_groups,
+        }
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+
+    def __repr__(self):
+        format_string = self.__class__.__name__ + ' ('
+        for i, sparse_args in enumerate(self.module_groups):
+            module = sparse_args['module']
+            format_string += '\n'
+            format_string += f'\tModule Group {i}\n'
+            format_string += f'\t    module: {module}\n'
+            for key in sorted(sparse_args.keys()):
+                if key == 'module':
+                    continue
+                format_string += f'\t    {key}: {sparse_args[key]}\n'
+        format_string += ')'
+        return format_string
+
+    def _pack_state(self):
+        state = defaultdict(dict)
+        for g in self.module_groups:
+            parametrization = g['module'].parametrizations['weight']
+            original_weight = parametrization.original
+            mask = None
+            # Find the mask in the FakeSparsity.
+            found = False
+            for p in parametrization:
+                if isinstance(p, FakeSparsity):
+                    parametrization = p
+                    found = True
+                    break
+            if found:
+                mask = parametrization.mask
+            state[original_weight]['mask'] = mask
+            state[original_weight]['fqn'] = g['fqn']
+            # Get all the tensors inside the module_group
+            state[original_weight].update(
```

Comment:
This doesnt look right. We cannot serialize the original weight as a key to the state.

Reply:
That what the optimizer does though -- I am mimicking what they do

Reply:
But I think using fqns as keys should be fine.